lib/eslint/eslint-helpers.js

```diff
@@ -151,94 +285,185 @@ async function globSearch({ basePath, patterns, configs }) {
-    const results = await Promise.all(
-        [...searches].map(
-            ([basePath, { patterns }]) => globSearch({ basePath, patterns, configs })
+    /*
+     * For convenience, we normalized the search map into an array of objects.
+     * Next, we filter out all searches that have no patterns. This happens
+     * primarily for the cwd, which is prepopulated in the searches map as an
+     * optimization. However, if it has no patterns, it means all patterns
+     * occur outside of the cwd and we can safely filter out that search.
+     */
+    const normalizedSearches = [...searches].map(
+        ([basePath, { patterns, rawPatterns }]) => ({ basePath, patterns, rawPatterns })
+    ).filter(({ patterns }) => patterns.length > 0);
+
+    const results = await Promise.allSettled(
+        normalizedSearches.map(
+            ({ basePath, patterns, rawPatterns }) => globSearch({
+                basePath,
+                patterns,
+                rawPatterns,
+                configs,
+                errorOnUnmatchedPattern
+            })
         )
     );
 
-    return [...new Set(results.flat())];
-}
-
-/**
- * Determines if a given glob pattern will return any results.
- * Used primarily to help with useful error messages.
- * @param {Object} options The options for the function.
- * @param {string} options.basePath The directory to search.
- * @param {string} options.pattern A glob pattern to match.
- * @returns {Promise<boolean>} True if there is a glob match, false if not.
- */
-function globMatch({ basePath, pattern }) {
-
-    let found = false;
-    const patternToUse = path.isAbsolute(pattern)
-        ? normalizeToPosix(path.relative(basePath, pattern))
-        : pattern;
+    const filePaths = [];
 
-    const matcher = new Minimatch(patternToUse);
+    for (let i = 0; i < results.length; i++) {
 
-    const fsWalkSettings = {
+        const result = results[i];
+        const currentSearch = normalizedSearches[i];
 
-        deepFilter(entry) {
-            const relativePath = normalizeToPosix(path.relative(basePath, entry.path));
+        if (result.status === "fulfilled") {
 
-            return !found && matcher.match(relativePath, true);
-        },
-
-        entryFilter(entry) {
-            if (found || entry.dirent.isDirectory()) {
-                return false;
+            // if the search was successful just add the results
+            if (result.value.length > 0) {
+                filePaths.push(...result.value);
+                continue;
             }
 
-            const relativePath = normalizeToPosix(path.relative(basePath, entry.path));
-
-            if (matcher.match(relativePath)) {
-                found = true;
-                return true;
-            }
-
-            return false;
+            // if the are no results we need to check for ignored files
+            await checkForIgnoredResults({
+                ...currentSearch,
+                errorOnUnmatchedPattern
+            });
```

Comment:
I think this should be removed.

With `errorOnUnmatchedPattern = true`, `globSearch` will either return some results or throw `UnmatchedSearchPatternsError`, so either way `globMultiSearch` will never fall into this branch?

With `errorOnUnmatchedPattern = false`, this causes a different behavior than that of eslintrc in regard to patterns that match only ignored files.

For example, run:

```
eslint "tools/*.js" --ignore-pattern "tools/*.js" --no-error-on-unmatched-pattern
```

In eslintrc mode, this exits with 0, meaning that `--no-error-on-unmatched-pattern` applies to patterns that match only ignored files too.

In flat config mode (current iteration of this PR), the same command throws an error:

```
Oops! Something went wrong! :(

ESLint: 8.26.0

You are linting "tools/*.js", but all of the files matching the glob pattern "tools/*.js" are ignored.
```

Reply:
> In eslintrc mode, this exits with 0, meaning that --no-error-on-unmatched-pattern applies to patterns that match only ignored files too.

Ah! I misunderstood what we were already doing. I thought —no-error-on-unmatched-pattern only applied to patterns without a match that weren’t caused by ignores. 

Really nice catch!